module/VuFind/src/VuFind/Auth/Shibboleth.php

```diff
@@ -295,4 +356,73 @@ protected function getRequiredAttributes()
 
         return $sortedUserAttributes;
     }
+
+    /**
+     * Add session id mapping to external_session table for single logout support
+     *
+     * @param \Laminas\Http\PhpEnvironment\Request $request Request object containing
+     * account credentials.
+     *
+     * @return \VuFind\Db\Row\User Object representing logged-in user.
+     */
+    protected function storeShibbolethSession($request)
+    {
+        $shib = $this->getConfig()->Shibboleth;
+        if (!isset($shib['session_id'])) {
+            return;
+        }
+        $shibSessionId = $this->getAttribute($request, $shib['session_id']);
+        if (null === $shibSessionId) {
+            return;
+        }
+        $localSessionId = $this->sessionManager->getId();
+        $externalSession = $this->getDbTableManager()->get('ExternalSession');
+        $externalSession->addSessionMapping($localSessionId, $shibSessionId);
+        $this->debug(
+            "Cached Shibboleth session id '$shibSessionId' for local session"
+            . " '$localSessionId'"
+        );
+    }
+
+    /**
+     * Fetch entityId used for authentication
+     *
+     * @param \Laminas\Http\PhpEnvironment\Request $request Request object
+     *
+     * @return string entityId of IdP
+     */
+    protected function getCurrentEntityId($request)
+    {
+        return $this->getAttribute($request, static::DEFAULT_IDPSERVERPARAM);
+    }
+
+    /**
+     * Extract attribute from request.
+     *
+     * @param \Laminas\Http\PhpEnvironment\Request $request   Request object
+     * @param string                               $attribute Attribute name
+     *
+     * @return string attribute value
+     */
+    protected function getAttribute($request, $attribute)
+    {
+        if ($this->useHeaders) {
+            $header = $request->getHeader($this->normalize($attribute));
+            return ($header) ? $header->getFieldValue() : null;
+        } else {
+            return $request->getServer()->get($attribute, null);
+        }
+    }
+
+    /**
+     * Convert name of server environment variable to header name.
+     *
+     * @param string $attribute name of server environment variable
+     *
+     * @return string header name
+     */
+    protected function normalize($attribute)
```

Comment:
This method might benefit from a more descriptive name, like `normalizeAttrName`, in case in future we need to normalize something else in a different way.